File: source-code/ninja-thief/episode-4/picosystem-micropython/player_ninja.py
from ninja import Ninja
import constants as Constants

class PlayerNinja(Ninja):

    def __init__(self, x, y):
        super().__init__(Ninja.Colour.BLUE, x, y)


    def update(self, dt, level_data):
        # If nothing is pressed, the player shouldn't move
        self.velocity_x = 0.0

        # Note: "else if" isn't used, because otherwise the sprite will still move when both buttons are pressed
        # Instead, we add/subtract the velocity, so if both are pressed, nothing happens
        if button(LEFT):
            self.velocity_x -= Constants.Player.MAX_SPEED
        
        if button(RIGHT):
            self.velocity_x += Constants.Player.MAX_SPEED

        # Handle climbing
        if self.can_climb:
            up = button(UP)
            down = button(DOWN)

            if up != down:
                # Only one of up and down are selected
                self.climbing_state = Ninja.ClimbingState.UP if up else Ninja.ClimbingState.DOWN
            
            elif self.climbing_state != Ninja.ClimbingState.NONE:
                # Player has already been climbing the ladder, and either none or both of up and down are pressed
                self.climbing_state = Ninja.ClimbingState.IDLE

        # Handle jumping
        # Note that we use the pressed function, which returns true if the button was just pressed (since the last frame)
        if pressed(A):
            if self.can_jump:
                self.jump(Constants.Player.JUMP_SPEED)

        super().update(dt, level_data)

File: source-code/ninja-thief/episode-4/32blit-lua/player_ninja.lua
require "ninja"

PlayerNinja = Ninja:new()

function PlayerNinja:new(x, y)
    return Ninja.new(PlayerNinja, Colour.BLUE, x, y)
end

function PlayerNinja:update(dt, level_data)
    -- If nothing is pressed, the player shouldn't move
    self.velocity_x = 0.0

    -- Note: "else if" isn't used, because otherwise the sprite will still move when both buttons are pressed
    -- Instead, we add/subtract the velocity, so if both are pressed, nothing happens
    if buttons.state & Button.LEFT ~= 0 then
        self.velocity_x = self.velocity_x - Constants.Player.MAX_SPEED
    end
    if buttons.state & Button.RIGHT ~= 0 then
        self.velocity_x = self.velocity_x + Constants.Player.MAX_SPEED
    end

    -- Handle climbing
    if self.can_climb then
        local up = buttons.state & Button.UP ~= 0
        local down = buttons.state & Button.DOWN ~= 0

        if up ~= down then
            -- Only one of up and down are selected
            self.climbing_state = up and ClimbingState.UP or ClimbingState.DOWN
        elseif self.climbing_state ~= ClimbingState.NONE then
            -- Player has already been climbing the ladder, and either none or both of up and down are pressed
            self.climbing_state = ClimbingState.IDLE
        end
    end

    -- Handle jumping
    -- Note that we use buttons.pressed, which only contains the buttons just pressed (since the last frame)
    if buttons.pressed & Button.A ~= 0 then
        if self.can_jump then
            self:jump(Constants.Player.JUMP_SPEED)
        end
    end

    Ninja.update(self, dt, level_data)
end
